File: app/src/main/java/com/WebSocketClient.java
// WebSocketClient.java
package com.gemweblive;

import android.content.Context;
import android.util.Base64;
import android.util.Log;

import androidx.annotation.NonNull;

import com.google.gson.FieldNamingPolicy;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.annotations.SerializedName;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.function.BiConsumer;
import java.util.function.Consumer;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.WebSocket;
import okhttp3.WebSocketListener;
import okhttp3.logging.HttpLoggingInterceptor;

public class WebSocketClient {

    private static final String TAG = "WebSocketClient";
    private static final String HOST = "generativelanguage.googleapis.com";
    private static final String API_KEY = "AIzaSyAIrTcT8shPcho-TFRI2tFJdCjl6_FAbO8"; // Use your actual API key
    private static final String SYSTEM_INSTRUCTION_TEXT = "You are a helpful assistant. Translate between English and Thai.\nBe direct and concise.";

    private final Context context;
    private final String model;
    private final int vadSilenceMs;
    private final String apiVersion;
    private final Runnable onOpen;
    private final Consumer<String> onMessage;
    private final BiConsumer<Integer, String> onClosing;
    private final Consumer<Throwable> onFailure;
    private final Runnable onSetupComplete;

    private WebSocket webSocket;
    private volatile boolean isSetupComplete = false;
    private volatile boolean isConnected = false;

    private final ExecutorService scope = Executors.newSingleThreadExecutor();
    private final Gson gson = new GsonBuilder()
            .disableHtmlEscaping()
            .setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
            .create();

    private PrintWriter logFileWriter;
    private final OkHttpClient client;

    public WebSocketClient(Context context, String model, int vadSilenceMs, String apiVersion,
                           Runnable onOpen, Consumer<String> onMessage, BiConsumer<Integer, String> onClosing,
                           Consumer<Throwable> onFailure, Runnable onSetupComplete) {
        this.context = context;
        this.model = model;
        this.vadSilenceMs = vadSilenceMs;
        this.apiVersion = apiVersion;
        this.onOpen = onOpen;
        this.onMessage = onMessage;
        this.onClosing = onClosing;
        this.onFailure = onFailure;
        this.onSetupComplete = onSetupComplete;

        HttpLoggingInterceptor loggingInterceptor = new HttpLoggingInterceptor(message -> {
            Log.d(TAG, message);
            if (logFileWriter != null) {
                logFileWriter.println(message);
            }
        });
        loggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);

        this.client = new OkHttpClient.Builder()
                .readTimeout(0, TimeUnit.MILLISECONDS)
                .pingInterval(30, TimeUnit.SECONDS)
                .addInterceptor(loggingInterceptor)
                .build();
    }

    public void connect() {
        if (isConnected) return;
        Log.i(TAG, "Attempting to connect...");

        try {
            File logDir = new File(context.getExternalFilesDir(null), "http_logs");
            if (!logDir.exists()) logDir.mkdirs();
            File logFile = new File(logDir, "network_log_" + System.currentTimeMillis() + ".txt");
            logFileWriter = new PrintWriter(new FileWriter(logFile, true), true);
            logFileWriter.println("--- New Session Log: " + new java.util.Date() + " ---");
        } catch (IOException e) {
            Log.e(TAG, "Failed to initialize log file", e);
            onFailure.accept(e);
            return;
        }

        String url = "wss://" + HOST + "/ws/google.ai.generativelanguage." + apiVersion + ".GenerativeService.BidiGenerateContent?key=" + API_KEY;
        Request request = new Request.Builder().url(url).build();

        webSocket = client.newWebSocket(request, new WebSocketListener() {
            @Override
            public void onOpen(@NonNull WebSocket webSocket, @NonNull Response response) {
                scope.execute(() -> {
                    Log.i(TAG, "WebSocket connection opened");
                    logFileWriter.println("WEB_SOCKET_OPENED");
                    isConnected = true;
                    sendConfigMessage();
                    onOpen.run();
                });
            }

            @Override
            public void onMessage(@NonNull WebSocket webSocket, @NonNull String text) {
                scope.execute(() -> {
                    Log.d(TAG, "Server message: " + text.substring(0, Math.min(text.length(), 500)) + "...");
                    logFileWriter.println("INCOMING MESSAGE: " + text);
                    try {
                        Map<?, ?> responseMap = gson.fromJson(text, Map.class);
                        if (responseMap.containsKey("setupComplete")) {
                            if (!isSetupComplete) {
                                isSetupComplete = true;
                                Log.i(TAG, "Server setup complete");
                                onSetupComplete.run();
                            }
                        } else {
                            WebSocketClient.this.onMessage.accept(text);
                        }
                    } catch (Exception e) {
                        Log.e(TAG, "Error parsing server message", e);
                        logFileWriter.println("ERROR PARSING INCOMING MESSAGE: " + e.getMessage());
                    }
                });
            }

            @Override
            public void onClosing(@NonNull WebSocket webSocket, int code, @NonNull String reason) {
                scope.execute(() -> {
                    Log.w(TAG, "WebSocket closing: " + code + " - " + reason);
                    logFileWriter.println("WEB_SOCKET_CLOSING: Code=" + code + ", Reason=" + reason);
                    cleanup();
                    WebSocketClient.this.onClosing.accept(code, reason);
                });
            }

            @Override
            public void onFailure(@NonNull WebSocket webSocket, @NonNull Throwable t, Response response) {
                scope.execute(() -> {
                    Log.e(TAG, "WebSocket failure", t);
                    logFileWriter.println("WEB_SOCKET_FAILURE: " + t.getMessage() + ", Response=" + (response != null ? response.code() : "null"));
                    cleanup();
                    WebSocketClient.this.onFailure.accept(t);
                });
            }
        });
    }

    private void sendConfigMessage() {
        try {
            SetupPayload payload = new SetupPayload(
                new Setup(
                    "models/" + model,
                    new GenerationConfig(),
                    new TranscriptionConfig(),
                    new TranscriptionConfig(),
                    new SystemInstruction(Collections.singletonList(new Part(SYSTEM_INSTRUCTION_TEXT))),
                    new RealtimeInputConfig(new AutomaticActivityDetection(vadSilenceMs))
                )
            );
            String configString = gson.toJson(payload);
            Log.d(TAG, "Sending config (length: " + configString.length() + "): " + configString);
            logFileWriter.println("OUTGOING CONFIG: " + configString);
            webSocket.send(configString);
        } catch (Exception e) {
            Log.e(TAG, "Failed to send config", e);
        }
    }
    
    public void sendAudio(byte[] audioData) {
        if (!isReady()) return;
        scope.execute(() -> {
            try {
                AudioPayload payload = new AudioPayload(
                    new RealtimeInput(
                        new Audio(Base64.encodeToString(audioData, Base64.NO_WRAP))
                    )
                );
                String messageToSend = gson.toJson(payload);
                Log.d(TAG, "Sending audio message (length: " + messageToSend.length() + ")");
                logFileWriter.println("OUTGOING AUDIO MESSAGE (length: " + messageToSend.length() + "): " + messageToSend.substring(0, Math.min(messageToSend.length(), 500)) + "...");
                webSocket.send(messageToSend);
            } catch (Exception e) {
                Log.e(TAG, "Failed to send audio", e);
                logFileWriter.println("ERROR SENDING AUDIO: " + e.getMessage());
            }
        });
    }

    public void disconnect() {
        scope.execute(this::cleanup);
    }

    private void cleanup() {
        if (isConnected) {
            Log.i(TAG, "Cleaning up WebSocket connection");
            if (webSocket != null) {
                webSocket.close(1000, "Normal closure");
                webSocket = null;
            }
        }
        if (logFileWriter != null) {
            logFileWriter.println("--- Session Log End ---");
            logFileWriter.close();
            logFileWriter = null;
        }
        isConnected = false;
        isSetupComplete = false;
    }

    public boolean isReady() {
        return isConnected && isSetupComplete;
    }

    // --- POJO Classes for GSON Serialization ---
    static class SetupPayload { Setup setup; public SetupPayload(Setup s) { this.setup = s; } }
    static class Setup {
        String model;
        GenerationConfig generationConfig;
        TranscriptionConfig inputAudioTranscription;
        TranscriptionConfig outputAudioTranscription;
        SystemInstruction systemInstruction;
        RealtimeInputConfig realtimeInputConfig;
        public Setup(String m, GenerationConfig gc, TranscriptionConfig iat, TranscriptionConfig oat, SystemInstruction si, RealtimeInputConfig ric) {
            this.model = m; this.generationConfig = gc; this.inputAudioTranscription = iat; this.outputAudioTranscription = oat; this.systemInstruction = si; this.realtimeInputConfig = ric;
        }
    }
    static class GenerationConfig { List<String> responseModalities = Collections.singletonList("AUDIO"); }
    static class TranscriptionConfig {} // Empty object
    static class SystemInstruction { List<Part> parts; public SystemInstruction(List<Part> p) { this.parts = p; } }
    static class Part { String text; public Part(String t) { this.text = t; } }
    static class RealtimeInputConfig { AutomaticActivityDetection automaticActivityDetection; public RealtimeInputConfig(AutomaticActivityDetection aad) { this.automaticActivityDetection = aad; } }
    static class AutomaticActivityDetection { int silenceDurationMs; public AutomaticActivityDetection(int s) { this.silenceDurationMs = s; } }

    static class AudioPayload { RealtimeInput realtimeInput; public AudioPayload(RealtimeInput r) { this.realtimeInput = r; } }
    static class RealtimeInput { Audio audio; public RealtimeInput(Audio a) { this.audio = a; } }
    static class Audio { String data; @SerializedName("mime_type") String mimeType = "audio/pcm;rate=16000"; public Audio(String d) { this.data = d; } }
}

File: app/src/main/java/com/WebSocketClient.kt
// app/src/main/java/com/gemweblive/WebSocketClient.kt
package com.gemweblive

import android.content.Context
import android.util.Base64
import android.util.Log
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.FieldNamingPolicy
import kotlinx.coroutines.*
import okhttp3.*
import okio.ByteString
import okhttp3.logging.HttpLoggingInterceptor
import java.io.File
import java.io.FileWriter
import java.io.PrintWriter
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

class WebSocketClient(
    private val context: Context,
    private val model: String,
    private val vadSilenceMs: Int,
    private val apiVersion: String,
    private val apiKey: String,
    private val sessionHandle: String?,
    private val onOpen: () -> Unit,
    private val onMessage: (String) -> Unit,
    private val onClosing: (Int, String) -> Unit,
    private val onFailure: (Throwable) -> Unit,
    private val onSetupComplete: () -> Unit
) {
    private var webSocket: WebSocket? = null
    @Volatile private var isSetupComplete = false
    @Volatile private var isConnected = false

    private val scope = CoroutineScope(Executors.newSingleThreadExecutor().asCoroutineDispatcher())
    private val gson: Gson = GsonBuilder()
        .disableHtmlEscaping()
        .setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
        .create()

    private var logFileWriter: PrintWriter? = null
    private lateinit var logFile: File

    private val client = OkHttpClient.Builder()
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .pingInterval(30, TimeUnit.SECONDS)
        .addInterceptor(HttpLoggingInterceptor(object : HttpLoggingInterceptor.Logger {
            override fun log(message: String) {
                Log.d(TAG, message)
                logFileWriter?.println(message)
            }
        }).apply {
            level = HttpLoggingInterceptor.Level.BODY
        })
        .build()

    companion object {
        private const val HOST = "generativelanguage.googleapis.com"
        private const val TAG = "WebSocketClient"

        // The system instruction text remains a single block here for readability
        private val SYSTEM_INSTRUCTION_TEXT = """
            ### **LLM System Prompt: Bilingual Live Thai-English Interpreter (Pattaya Bar Scene)**

            **1. ROLE AND OBJECTIVE**

            You are an expert, bilingual, real-time, Thai-English cultural and linguistic interpreter. Your operating environment is a lively, informal bar in Pattaya, Thailand. Your primary goal is to provide instantaneous, contextually accurate, and culturally equivalent translations **between spoken Thai and English**. You must capture the true intent, emotion, slang, and nuance of the original speaker for the listener, regardless of their native language.

            **2. CORE PRINCIPLES**

            * **Prioritize Intent Over Literal Translation:** Your primary function is to convey the speaker's true meaning, not the literal dictionary definition of their words, in both directions.
            * **Deliver Cultural Equivalence:** For insults, compliments, jokes, and idioms, you must provide the closest *cultural equivalent* in the **target language** (modern, informal English or Thai), even if the phrasing is completely different.
            * **Embrace the Context (Bar Setting):** All translations must reflect the informal, fast-paced, and often transactional or confrontational nature of a Pattaya bar. This applies whether the original speaker is Thai or a foreigner. Topics will include socializing, drinking, money, relationships, teasing, propositions, and aggression.
            * **Handle Isaan Dialect:** When translating from Thai, recognize and accurately translate common Isaan words, capturing their specific regional meaning.
            * **Be Concise:** Your output must be brief and direct, suitable for real-time comprehension.

            **3. STRICT LANGUAGE CONSTRAINT PROTOCOL**

            * **ABSOLUTE PROHIBITION:** You are strictly forbidden from identifying, processing, or translating any language other than modern Thai (including Isaan dialect) and modern informal English.
            * ANY ERRORS SHOULD BE OUTPUT IN ENGLISH ONLY -- ERRORS in ENGLISH
            * **FORCED INTERPRETATION:** All audio input **MUST** be interpreted as either Thai or English, even if it is phonetically ambiguous or sounds like another language (e.g., Mandarin, Cantonese). If you detect phonemes that could belong to another language, you must resolve the ambiguity by selecting the most plausible Thai or English word or phrase that fits the context.
            * **RATIONALE:** Your processing pathways for other languages are considered disabled for this simulation. Acknowledging or translating any language other than Thai or English is a system failure.
        """.trimIndent()
    }

    private fun sendConfigMessage() {
        try {
            // --- NEW: Split the instruction text into paragraphs ---
            val instructionParts = SYSTEM_INSTRUCTION_TEXT.split(Regex("\n\n+")).map {
                mapOf("text" to it.trim())
            }

            val setupConfig = mutableMapOf<String, Any>(
                "model" to "models/$model",
                "generationConfig" to mapOf(
                    "responseModalities" to listOf("AUDIO")
                ),
                "inputAudioTranscription" to emptyMap<String, Any>(),
                "outputAudioTranscription" to emptyMap<String, Any>(),
                // --- MODIFIED: Use the new list of parts ---
                "systemInstruction" to mapOf("parts" to instructionParts),
                "realtimeInputConfig" to mapOf(
                    "automaticActivityDetection" to mapOf(
                        "silenceDurationMs" to vadSilenceMs
                    )
                ),
                "contextWindowCompression" to mapOf(
                    "slidingWindow" to emptyMap<String, Any>()
                )
            )

            val sessionResumption = if (sessionHandle != null) {
                mapOf("handle" to sessionHandle)
            } else {
                emptyMap<String, Any>()
            }
            setupConfig["sessionResumption"] = sessionResumption

            val fullConfig = mapOf("setup" to setupConfig)
            val configString = gson.toJson(fullConfig)
            webSocket?.send(configString)

        } catch (e: Exception) {
            Log.e(TAG, "Failed to send config", e)
        }
    }

    fun connect() {
        if (isConnected) return
        Log.i(TAG, "Attempting to connect with session handle: ${sessionHandle?.take(10)}...")

        try {
            val logDir = File(context.getExternalFilesDir(null), "websocket_logs")
            if (!logDir.exists()) logDir.mkdirs()
            logFile = File(logDir, "session_log_${System.currentTimeMillis()}.txt")
            logFileWriter = PrintWriter(FileWriter(logFile, true), true)
            logFileWriter?.println("--- New WebSocket Session Log: ${java.util.Date()} ---")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to initialize log file", e)
            onFailure(e)
            return
        }

        val request = Request.Builder()
            .url("wss://$HOST/ws/google.ai.generativelanguage.$apiVersion.GenerativeService.BidiGenerateContent?key=$apiKey")
            .build()

        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                scope.launch {
                    isConnected = true
                    logFileWriter?.println("WEB_SOCKET_OPENED (HTTP Status: ${response.code})")
                    sendConfigMessage()
                    onOpen()
                }
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                scope.launch {
                    logFileWriter?.println("INCOMING TEXT FRAME: $text")
                    processIncomingMessage(text)
                }
            }

            override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
                scope.launch {
                    val decodedText = bytes.utf8()
                    Log.d(TAG, "INCOMING BINARY FRAME (decoded): ${decodedText.take(200)}...")
                    logFileWriter?.println("INCOMING BINARY FRAME (decoded): $decodedText")
                    processIncomingMessage(decodedText)
                }
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                scope.launch {
                    logFileWriter?.println("WEB_SOCKET_CLOSING: Code=$code, Reason=$reason")
                    cleanup()
                    this@WebSocketClient.onClosing(code, reason)
                }
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                scope.launch {
                    Log.e(TAG, "WebSocket failure", t)
                    logFileWriter?.println("WEB_SOCKET_FAILURE: ${t.message}")
                    cleanup()
                    this@WebSocketClient.onFailure(t)
                }
            }
        })
    }
    
    private fun processIncomingMessage(messageText: String) {
        try {
            if (messageText.contains("\"setupComplete\"")) {
                if (!isSetupComplete) {
                    isSetupComplete = true
                    Log.i(TAG, "Server setup complete.")
                    onSetupComplete()
                }
            } else {
                this@WebSocketClient.onMessage(messageText)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error processing incoming message: '$messageText'", e)
        }
    }

    fun sendAudio(audioData: ByteArray) {
        if (!isReady()) return
        scope.launch {
            try {
                val base64Audio = Base64.encodeToString(audioData, Base64.NO_WRAP)
                val realtimeInput = mapOf(
                    "realtimeInput" to mapOf(
                        "audio" to mapOf(
                            "data" to base64Audio,
                            "mime_type" to "audio/pcm;rate=16000"
                        )
                    )
                )
                val messageToSend = gson.toJson(realtimeInput)
                webSocket?.send(messageToSend)
            } catch (e: Exception) {
                Log.e(TAG, "Failed to send audio", e)
            }
        }
    }

    fun disconnect() {
        scope.launch {
            cleanup()
        }
    }

    private fun cleanup() {
        if (isConnected) {
            webSocket?.close(1000, "Normal closure")
            webSocket = null
        }
        logFileWriter?.close()
        logFileWriter = null
        isConnected = false
        isSetupComplete = false
    }

    fun isReady(): Boolean = isConnected && isSetupComplete
    fun isConnected(): Boolean = isConnected
}
